### PR TITLE
SWATCH-2304: Add a metric for usage aggregated per hour

### PR DESCRIPTION
**Pull request by Sgitario**
Jira issue: SWATCH-2304

## Description
When usage is aggregated, it should increment a counter named "swatch_billable_usage_total" with the billable amount in metric units, having labels:

product (swatch product tag)
metric_id (metric ID)
billing_provider (aws/azure)
status (aggregated)

## Testing
1. podman-compose up
2. setup database schema: `./gradlew liquibaseUpdate`
3. start the tally worker: `RHSM_SUBSCRIPTIONS_ENABLE_SYNCHRONOUS_OPERATIONS=true DEV_MODE=true PROM_URL=http://localhost:8082/api/metrics/v1/telemeter/api/v1 SWATCH_CONTRACTS_INTERNAL_SERVICE=http://localhost:8003 SERVER_PORT=8000 SPRING_PROFILES_ACTIVE=worker,kafka-queue,api ./gradlew clean :bootRun`
4.- start the contracts service: `SERVER_PORT=8003 QUARKUS_MANAGEMENT_PORT=9003 ./gradlew :swatch-contracts:quarkusDev`
5.- start the azure service: `SERVER_PORT=8002 QUARKUS_MANAGEMENT_PORT=9002 ./gradlew :swatch-producer-azure:quarkusDev`
6.- start the billable usage service: `SERVER_PORT=8001 QUARKUS_MANAGEMENT_PORT=9001 KSTREAM_BILLABLE_USAGE_AGGREGATION_WINDOW_DURATION=10s KSTREAM_BILLABLE_USAGE_AGGREGATION_GRACE_DURATION=2s ./gradlew :swatch-billable-usage:quarkusDev`

Note that we need all the above services to ensure all the Kafka topics are created automatically.

7. insert test data:
```
insert into offering(sku) values ('MW02393');
insert into sku_product_tag(sku,product_tag) values ('MW02393','ansible-aap-managed');
INSERT INTO contracts VALUES ('b3a5e8b7-6260-48f4-a9eb-63a584763ce9', '12914277', now(), '2024-08-01 01:01:01.717275+00', NULL, 'org123', 'MW02393', 'aws', 'aws123', 'ansible-aap-managed', 'bcwuzc8ww10vvxfair55mliy8');
```

8. create a few messages in `platform.rhsm-subscriptions.tally`:
1:
```
{"org_id": "org123", "tally_snapshots": [{"billing_provider":"aws", "billing_account_id": "aws123", "snapshot_date": "2024-08-10T01:10:28Z", "product_id": "ansible-aap-managed", "sla": "Premium", "usage": "Production", "granularity": "Hourly", "tally_measurements": [{"hardware_measurement_type": "AWS", "metric_id": "Managed-nodes", "value": 12, "currentTotal": 100}]}]}
``` 
2:
```
{"org_id": "org123", "tally_snapshots": [{"billing_provider":"aws", "billing_account_id": "123", "snapshot_date": "2023-12-21T01:10:28Z", "product_id": "rosa", "sla": "Premium", "usage": "Production", "granularity": "Hourly", "tally_measurements": [{"hardware_measurement_type": "AWS", "metric_id": "Cores", "value": 12, "currentTotal": 100}]}]}
```
3:
```
{"org_id": "org123", "tally_snapshots": [{"billing_provider":"aws", "billing_account_id": "123", "snapshot_date": "2023-12-21T01:10:28Z", "product_id": "rosa", "sla": "Premium", "usage": "Production", "granularity": "Hourly", "tally_measurements": [{"hardware_measurement_type": "AWS", "metric_id": "Cores", "value": 12, "currentTotal": 100}]}]}
```

9. verification: 

The metrics from the swatch billable usage service should contain the new metric:

```
curl http://localhost:9001/metrics  | grep swatch_
# TYPE rhsm_subscriptions_swatch_billable_usage counter
# HELP rhsm_subscriptions_swatch_billable_usage  
rhsm_subscriptions_swatch_billable_usage_total{billing_provider="aws",metric_id="Cores",product="rosa"} 25.0
```

### CODE REVIEW

File: swatch-billable-usage/src/main/java/com/redhat/swatch/billable/usage/kafka/streams/StreamTopologyProducer.java [hunk 1]
```diff
@@ -89,9 +92,31 @@ public Topology buildTopology() {
...
+
+      if (aggregate.getStatus() != null) {
+        tags.addAll(List.of("status", aggregate.getStatus().toString()));
+      }
+
+      meterRegistry
+          .counter(USAGE_TOTAL_METRIC, tags.toArray(new String[0]))
+          .increment(aggregate.getTotalValue().doubleValue());
+    }
+  }
```
[the remark this answers is not in the document]
Comment (kartikshahc) [reply swatch-billable-usage/src/main/java/com/redhat/swatch/billable/usage/kafka/streams/StreamTopologyProducer.java:121]: Both solutions work, but I agree that the suggested approach is more readable. Additionally, I noticed that the status tag is not being populated, as we’re currently adding tags only when consuming the message. Should we take any additional steps after persisting to the database to ensure aggregation is shown properly by status?

**Review by kartikshahc — CHANGES_REQUESTED** [review]

File: swatch-billable-usage/src/main/java/com/redhat/swatch/billable/usage/kafka/streams/StreamTopologyProducer.java [hunk 2]
```diff
@@ -35,23 +39,22 @@
...
 import org.candlepin.subscriptions.billable.usage.BillableUsage;
 import org.candlepin.subscriptions.billable.usage.BillableUsageAggregate;
 import org.candlepin.subscriptions.billable.usage.BillableUsageAggregateKey;
 
+@AllArgsConstructor
 @ApplicationScoped
 @Slf4j
 public class StreamTopologyProducer {
 
+  public static final String USAGE_TOTAL_METRIC = "rhsm-subscriptions.swatch_billable_usage_total";
```
[the remark this answers is not in the document]
Comment (Sgitario) [reply swatch-billable-usage/src/main/java/com/redhat/swatch/billable/usage/kafka/streams/StreamTopologyProducer.java:53]: I followed the same convention than in https://github.com/RedHatInsights/rhsm-subscriptions/pull/3888/files#diff-c6d3287c0093e63d019e8d7a596128b872c21194e95ac5a29bd5766aa98c194cR63. 
I have no objections to simply name it "swatch_billable_usage_total", but all the metrics should follow the same convention.

[the remark this answers is not in the document]
Comment (Sgitario) [reply swatch-billable-usage/src/main/java/com/redhat/swatch/billable/usage/kafka/streams/StreamTopologyProducer.java:53]: According to the document linked to the epic issue, it should be "swatch_billable_usage_total". Updated.
Will create another pull request to also update the metric from the above other pull request.

File: swatch-billable-usage/src/main/java/com/redhat/swatch/billable/usage/kafka/streams/StreamTopologyProducer.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (Sgitario) [reply swatch-billable-usage/src/main/java/com/redhat/swatch/billable/usage/kafka/streams/StreamTopologyProducer.java]: Using the builder looks much better, thanks for the suggestion @awood 
@kartikshahc there will be other metrics that will capture the status after persisting.

**Comment by Sgitario** [comment]
@awood @kartikshahc PR updated, mind taking another look?

File: swatch-billable-usage/src/main/java/com/redhat/swatch/billable/usage/kafka/streams/StreamTopologyProducer.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (Sgitario) [reply swatch-billable-usage/src/main/java/com/redhat/swatch/billable/usage/kafka/streams/StreamTopologyProducer.java]: https://github.com/RedHatInsights/rhsm-subscriptions/pull/3961

**Review by awood — APPROVED** [review]
Looks good!

**Review by kartikshahc — APPROVED** [review]
LGTM

**Comment by Aurobinda55** [comment]
Verification steps -

- Added a contract for Rosa 
- Created two events for Rosa 
- Sync hourly tally 
- Checked the billable usage service pod and got the below metric usage 
```
# HELP swatch_billable_usage  
swatch_billable_usage_total{billing_provider="aws",metric_id="INSTANCE_HOURS",product="rosa"} 99.0
swatch_billable_usage_total{billing_provider="aws",metric_id="CORES",product="rosa"} 48.0
```
Looks good to me. The failed tests are unrelated to this PR. I deployed one EE against this PR and ran the failed test manually and all the failed tests worked fine on my deployed EE. SO, good to merge this PR.

**Pull request merged**